ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java

```diff
@@ -333,20 +333,34 @@ private EndpointResponse handleQuery(final KsqlSecurityContext securityContext,
                 metrics.recordStatusCode(statusCode);
                 metrics.recordRequestSize(requestBytes);
                 final PullQueryResult r = resultForMetrics.get();
-                final PullSourceType sourceType = Optional.ofNullable(r).map(
-                    PullQueryResult::getSourceType).orElse(PullSourceType.UNKNOWN);
-                final PullPhysicalPlanType planType = Optional.ofNullable(r).map(
-                    PullQueryResult::getPlanType).orElse(PullPhysicalPlanType.UNKNOWN);
-                final RoutingNodeType routingNodeType = Optional.ofNullable(r).map(
-                    PullQueryResult::getRoutingNodeType).orElse(RoutingNodeType.UNKNOWN);
-                metrics.recordResponseSize(responseBytes, sourceType, planType, routingNodeType);
-                metrics.recordLatency(startTimeNanos, sourceType, planType, routingNodeType);
-                metrics.recordRowsReturned(
-                    Optional.ofNullable(r).map(PullQueryResult::getTotalRowsReturned).orElse(0L),
-                    sourceType, planType, routingNodeType);
-                metrics.recordRowsProcessed(
-                    Optional.ofNullable(r).map(PullQueryResult::getTotalRowsProcessed).orElse(0L),
-                    sourceType, planType, routingNodeType);
+                if (r == null) {
+                  metrics.recordResponseSizeForError(responseBytes);
+                  metrics.recordLatencyForError(startTimeNanos);
```

Comment:
Can use `recordErrorMetrics`?

Reply:
Good eye! That's a method in PullQueryPublisher, so we can't use it here. I did think about pulling these common sets of metric-recording operations into the PullQueryMetrics class, though.